src/widgets/index/index.ts

```diff
@@ -342,6 +344,105 @@ const index = (props: IndexProps): Index => {
         // it at the index level because it's either: all of them or none of them
         // that are stalled. The queries are performed into a single network request.
         instantSearchInstance.scheduleStalledRender();
+
+        if (__DEV__) {
+          type StateToWidgets = {
+            [TParameter in keyof IndexUiState]: Array<Widget['$$type']>;
+          };
+
+          const stateToWidgetsMap: StateToWidgets = {
+            query: ['ais.searchBox', 'ais.autocomplete', 'ais.voiceSearch'],
+            refinementList: ['ais.refinementList'],
+            menu: ['ais.menu'],
+            hierarchicalMenu: ['ais.hierarchicalMenu'],
+            numericMenu: ['ais.numericMenu'],
+            ratingMenu: ['ais.ratingMenu'],
+            range: ['ais.range'],
+            toggle: ['ais.toggleRefinement'],
+            geoSearch: ['ais.geoSearch'],
+            sortBy: ['ais.sortBy'],
+            page: ['ais.pagination', 'ais.infiniteHits'],
+            hitsPerPage: ['ais.hitsPerPage'],
+            configure: ['ais.configure'],
+          };
+
+          const mountedWidgets = this.getWidgets()
+            .map(widget => widget.$$type)
+            .filter(Boolean);
+
+          const missingWidgets = Object.keys(localUiState).reduce<
+            Array<Partial<StateToWidgets>>
+          >((acc, parameter) => {
+            const requiredWidgets: Array<Widget['$$type']> =
+              stateToWidgetsMap[parameter];
+
+            if (
+              requiredWidgets.every(
```

Comment:
perf doesn't really matter in dev, but you can swap this condition around and it will be triggered as soon as something is off:

```js
if (requiredWidgets.some(requiredWidget => mountedWidgets.includes(requiredWidget) ===  false) === false)
```

This looks somewhat more ugly, which is why i  used ===  false, instead of !

Reply:
This can actually be simplified to this:

```ts
!requiredWidgets.some(requiredWidget =>
  mountedWidgets.includes(requiredWidget)
)
```